internal/controller/driver_controller.go

```diff
@@ -18,45 +18,401 @@ package controller
 	Scheme *runtime.Scheme
 }
 
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+// A local reconcile object tied to a single reconcile iteration
+type driverReconcile struct {
+	DriverReconciler
+
+	ctx        context.Context
+	log        logr.Logger
+	driver     csiv1a1.Driver
+	driverName string
+	driverType string
+	images     map[string]string
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DriverReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	enqueueFromOwnerRefAnnotation := handler.EnqueueRequestsFromMapFunc(
+		func(_ context.Context, obj client.Object) []reconcile.Request {
+			ownerRef := obj.GetAnnotations()[ownerRefAnnotationKey]
+			ownerObjKey := client.ObjectKey{}
+			if err := json.Unmarshal([]byte(ownerRef), &ownerObjKey); err != nil {
+				return nil
+			}
+
+			return []reconcile.Request{{
+				NamespacedName: types.NamespacedName{
+					Name:      ownerObjKey.Name,
+					Namespace: ownerObjKey.Namespace,
+				},
+			}}
+		},
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1a1.Driver{}).
+		Watches(&storagev1.CSIDriver{}, enqueueFromOwnerRefAnnotation).
+		Complete(r)
+}
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the Driver object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
 func (r *DriverReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	reconcileHandler := driverReconcile{}
+	reconcileHandler.DriverReconciler = *r
+	reconcileHandler.ctx = ctx
+	reconcileHandler.log = ctrllog.FromContext(ctx)
+	reconcileHandler.driver.Name = req.Name
+	reconcileHandler.driver.Namespace = req.Namespace
+
+	return reconcileHandler.reconcile()
+}
+
+func (r *driverReconcile) reconcile() (ctrl.Result, error) {
+	r.log.Info("Enter Reconcile", "req", client.ObjectKeyFromObject(&r.driver))
+
+	// Load the driver desired state based on driver resource, operator config resource and default values.
+	if err := r.LoadAndValidateDesiredState(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Concurrently reconcile different aspects of the clusters actual state to meet
+	// the desired state defined on the driver object
+	errChan := utils.RunConcurrently(
+		r.reconcileK8sCsiDriver,
+		r.reconcilePluginDeamonSet,
+		r.reconcileProvisionerDeployment,
+		r.reconcileLivnessService,
+	)
```

Comment:
@nb-ohad what is the main reason to have to run with Concurrently?

Reply:
Two reasons:
1. I dislike the notion that reconcile steps that have nothing to do with each other are blocked on the first one that fails. Take to following example:

Let's say that to get the cluster to the desired state we need to update both resource A and resource B. With the way we traditionally wrote reconcilers, we will have code that calls either `reconcileA()` then `reconcileB()` or `reconcileB()` then `reconcileA()`. 

If there aren't any dependencies between the reconciliation of A and B, the way we do it right now is sub-optimal. Because, by definition, a reconcile iteration should try to get the cluster as close as possible to the desired state but if we cut the loop after the first failure the second reconcile function will never be called. 

Using concurrency makes sure that all steps are running concurrently and errors are aggregated so everything that can be reconciled in this iteration will be reconciled before acting on any requeue or error. 

With that said, the code guarantees that we will not ignore any requeuing of events, so we are safe from that regard.

Reply:
"We still write them down the exact way we did before." **_plus_** keeping in mind that we need to code each step w/o data races (race conditions are fine though) and I totally agree/kinda make a hard requirement for "good set of unit tests"

However, I want to zoom in to the context a bit, I like parallelism among different drivers but not concurrency b/n different components in same driver as all steps in same driver should be successful for a meaningful work to be done, like a PVC creation (controller) and bounding (node) to workload.

I'm not opposing the idea however a little doubtful w/ the outcome. Open to be corrected.

Reply:
Ack, I'm not against the concurrency and only concerned about the outcome and maintenance.

Nevertheless, it is easy to switch if we want to forgo or improve it. can be resolved.